compiler.go

```diff
@@ -594,19 +601,37 @@ func (t *task) link(parseRes parser.Result, deps linker.Files) (linker.File, err
 	return file, nil
 }
 
+func needsSourceInfo(parseRes parser.Result, mode SourceInfoMode) bool {
+	return mode != SourceInfoNone && parseRes.AST() != nil && parseRes.FileDescriptorProto().SourceCodeInfo == nil
+}
+
 func (t *task) asParseResult(name string, r SearchResult) (parser.Result, error) {
 	if r.ParseResult != nil {
 		if r.ParseResult.FileDescriptorProto().GetName() != name {
 			return nil, fmt.Errorf("search result for %q returned descriptor for %q", name, r.ParseResult.FileDescriptorProto().GetName())
 		}
-		return r.ParseResult, nil
+		res := r.ParseResult
+		if fileNeedsResolution(res.FileDescriptorProto()) || needsSourceInfo(res, t.e.c.SourceInfoMode) {
+			// If the file descriptor needs linking, it will be mutated during the
+			// next stage. So to make the mutations thread-safe, we must make a
+			// defensive copy.
```

Comment:
Not sure if this is practical, but I wonder if having some wrapper around `descriptorpb` types to separate them into read/write views might make this safer. Basically have a Reader view which is always safe to call, and have an accessor to get a writer which does a sync.Once() to clone the object before performing any mutation.

I just worry about having to know in advance all the possible mutations you might perform on an object before they are made.

Reply:
I'm trying to image how we could do that without creating an entirely parallel structure (at that point it would be no cheaper than just _always_ cloning).

I guess the other approach is to _always_ clone and not try to be clever about it just to save some cycles.

You know, that's probably what we should do. That's what I did in the previous compiler (jhump/protoreflect/desc/protoparse): https://github.com/jhump/protoreflect/blob/0facb742c9dcbc99667de77a4b74b842f9aa0bb3/desc/protoparse/parser.go#L532-L535

So I'll update this PR so cloning is unconditional. That should simplify it a bit.

Reply:
I think that's probably best. If a user wants ultimate performance, they could first compile shared resources like the descriptor.proto file up front, and then provide the fully linked protoreflect.FileDescriptor to the resolver on subsequent calls.